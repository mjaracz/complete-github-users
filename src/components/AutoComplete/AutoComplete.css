.autoComplete {
    display: flex;
    justify-content: center;
    align-items: center;
    flex-wrap: wrap;
    width: 180px;
    height: auto;
}

.autoComplete__textField {
    position: relative;
    display: inline-block;
    font-family: 'Ubuntu', sans-serif;
    font-size: 16px;
    line-height: 1.5;
    overflow: hidden;
    width: 180px;
    margin-bottom: 12px;
}

.autoComplete__textField > input,
.autoComplete__textField > textarea {
    display: block;
    box-sizing: border-box;
    margin: 0;
    border: none;
    border-top: solid 27px transparent;
    border-bottom: solid 1px var(--secondary-color);
    padding: 5px;
    width: 100%;
    height: inherit;
    color: var(--secondary-color);
    background-color: transparent;
    box-shadow: none;
    font-family: 'Ubuntu', sans-serif;
    font-size: inherit;
    line-height: inherit;
    caret-color: var(--secondary-color);
    transition: border-bottom 0.2s, background-color 0.2s;
}

.autoComplete__textField > input + span,
.autoComplete__textField > textarea + span {
    position: absolute;
    top: 0;
    left: 0;
    right: 0;
    bottom: 0;
    display: block;
    box-sizing: border-box;
    color: var(--secondary-color);
    font-size: 75%;
    line-height: 18px;
    pointer-events: none;
    font-family: 'Ubuntu', sans-serif;
    padding: 6px;
    transition: .3s;
}

.autoComplete__textField > input + span::after,
.autoComplete__textField > textarea + span::after {
    content: "";
    position: absolute;
    left: 0;
    bottom: 0;
    display: block;
    width: 100%;
    height: 2px;
    background-color: var(--secondary-hover-color);
    transform-origin: bottom center;
    transform: scaleX(0);
    transition: transform 0.2s;
}

.autoComplete__textField > input:hover,
.autoComplete__textField > textarea:hover {
    border-bottom-color: var(--secondary-hover-color);
}

.autoComplete__textField > input:not(:focus):placeholder-shown + span,
.autoComplete__textField > textarea:not(:focus):placeholder-shown + span {
    font-size: inherit;
    line-height: 56px;
}

.autoComplete__textField > input:focus,
.autoComplete__textField > textarea:focus {
    outline: none;
    color: var(--secondary-hover-color);
}

.autoComplete__textField > input:focus + span,
.autoComplete__textField > textarea:focus + span {
    color: var(--secondary-hover-color);
    border: 2px solid var(--secondary-hover-color);
    border-radius: 2px;
}

.autoComplete__textField > input:focus + span::before,
.autoComplete__textField > textarea:focus + span::before {
    opacity: 0.12;
}

.autoComplete__textField > input:focus + span::after,
.autoComplete__textField > textarea:focus + span::after {
    transform: scale(1, 0.2);
}

.autoComplete__textField > input:disabled,
.autoComplete__textField > textarea:disabled {
    border-bottom-color: var(--primary-color);
    color: var(--secondary-hover-color);
}

.autoComplete__textField > input:disabled + span,
.autoComplete__textField > textarea:disabled + span {
    color: var(--primary-color);
}

@media not all and (min-resolution:.001dpcm) {
    @supports (-webkit-appearance:none) {
        .autoComplete__textField > input,
        .autoComplete__textField > input + span,
        .autoComplete__textField > input + span::after,
        .autoComplete__textField > textarea,
        .autoComplete__textField > textarea + span,
        .autoComplete__textField > textarea + span::after {
            transition-duration: 0.1s;
        }
    }
}

.autoComplete__optionsList {
    transition: .4s;
    position: relative;
    display: block;
    width: 180px;
    max-height: 250px;
    padding: 0;
    margin: 0;
    background-color: var(--secondary-color);
    border-radius: 2px;
}

.optionsList__item {
    width: 180px;
    height: 40px;
    display: flex;
    justify-content: center;
    align-items: center;
    border-bottom: 2px dashed var(--primary-color);
    color: var(--background-color);
    font-family: 'Ubuntu', sans-serif;
    letter-spacing: 2px;
    text-transform: full-size-kana;
}

.optionsList__item:last-child {
    border-bottom: none;
}