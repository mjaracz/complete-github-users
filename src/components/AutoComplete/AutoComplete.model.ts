export interface AutoCompleteProps<Option, Key extends keyof Option> {
  options: Option[];
  optionKey: Key;
}
