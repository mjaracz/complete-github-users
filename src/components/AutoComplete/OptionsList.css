.autoComplete__optionsList {
    opacity: 0;
    visibility: hidden;
    position: relative;
    width: 180px;
    max-height: 250px;
    padding: 0;
    margin: 0;
    background-color: var(--secondary-color);
    border-radius: 2px;
    transition: visibility .2s ease-in-out, opacity .2s ease-in-out;
    overflow: hidden;
}
.autoComplete__optionsList--show {
    visibility: visible;
    opacity: 1;
}

.optionsList__item {
    width: 180px;
    height: 40px;
    display: flex;
    justify-content: center;
    align-items: center;
    border-bottom: 2px dashed var(--primary-color);
    color: var(--background-color);
    font-family: 'Ubuntu', sans-serif;
    font-size: 16px;
    letter-spacing: 2px;
    cursor: pointer;
}

.optionsList__item:last-child {
    border-bottom: none;
}