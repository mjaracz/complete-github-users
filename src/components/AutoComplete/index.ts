export * from './AutoComplete';
export * from './AutoComplete.model';
