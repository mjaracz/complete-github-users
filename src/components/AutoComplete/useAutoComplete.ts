import {
  createRef,
  useCallback,
  useEffect,
  useState,
  MouseEvent,
  ChangeEvent,
} from 'react';

export const useAutoComplete = <Option>(
  propsOptions: Option[],
  optionKey: keyof Option
) => {
  const [showOptions, setShowOptions] = useState<boolean>(false);
  const [options, setOptions] = useState<Option[]>(propsOptions);
  const [label, setLabel] = useState<string>();
  const refAutoComplete = createRef<HTMLDivElement>();

  const hiddenOptionsList = useCallback(
    (e) => {
      const allowToHide =
        refAutoComplete && !refAutoComplete.current?.contains(e.currentTarget);
      if (allowToHide) {
        setShowOptions(false);
      }
    },
    [refAutoComplete]
  );

  useEffect(() => {
    document.addEventListener('mousedown', hiddenOptionsList);
    return function cleanup() {
      document.removeEventListener('mousedown', hiddenOptionsList);
    };
  }, [hiddenOptionsList]);

  const focusTextField = () => {
    setShowOptions(true);
  };
  const filterOptions = (e: ChangeEvent<HTMLInputElement>) => {
    setLabel(e.currentTarget.value);
    setOptions(
      propsOptions.filter((option) =>
        `${option[optionKey]}`.includes(e.currentTarget.value)
      )
    );
  };
  const onClickOption = (e: MouseEvent<HTMLLIElement>) => {
    setLabel(e.currentTarget.innerText);
  };

  return {
    label,
    showOptions,
    options,
    refAutoComplete,
    focusTextField,
    filterOptions,
    onClickOption,
  };
};
