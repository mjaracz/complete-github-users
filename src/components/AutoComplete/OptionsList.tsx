import React, { MouseEvent } from 'react';
import './OptionsList.css';
import clsx from 'clsx';

export interface OptionsListProps<Option> {
  showOptions: boolean;
  options: Option[];
  optionKey: keyof Option;
  onClickOption: (e: MouseEvent<HTMLLIElement>) => void;
}

export function OptionsList<Option>({
  showOptions,
  options,
  optionKey,
  onClickOption,
}: OptionsListProps<Option>) {
  return (
    <ul
      data-testid="autoComplete__optionsList"
      className={clsx('autoComplete__optionsList', {
        'autoComplete__optionsList--show': showOptions,
      })}
    >
      {options.map((option, index) => (
        <li
          data-testid="optionsList__item"
          key={`-${index}`}
          className="optionsList__item"
          onClick={onClickOption}
        >
          {option[optionKey]}
        </li>
      ))}
    </ul>
  );
}
