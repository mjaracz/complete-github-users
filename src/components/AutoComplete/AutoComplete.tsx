import React from 'react';
import { AutoCompleteProps } from './AutoComplete.model';
import { OptionsList } from './OptionsList';
import { useAutoComplete } from './useAutoComplete';
import './AutoComplete.css';

export function AutoComplete<Option>(
  props: AutoCompleteProps<Option, keyof Option>
) {
  const {
    label,
    showOptions,
    options,
    refAutoComplete,
    focusTextField,
    filterOptions,
    onClickOption,
  } = useAutoComplete<Option>(props.options, props.optionKey);

  return (
    <div
      data-testid="autoComplete"
      ref={refAutoComplete}
      className="autoComplete"
    >
      <label
        data-testid="autoComplete__textField"
        onClick={focusTextField}
        className="autoComplete__textField"
      >
        <input onChange={filterOptions} value={label} placeholder=" " />
        <span>provide username</span>
      </label>
      <OptionsList
        options={options}
        optionKey={props.optionKey}
        showOptions={showOptions}
        onClickOption={onClickOption}
      />
    </div>
  );
}
