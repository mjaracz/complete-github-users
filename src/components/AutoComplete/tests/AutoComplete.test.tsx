import { AutoComplete } from '../AutoComplete';
import { render } from '@testing-library/react';

describe('<AutoComplete/>', () => {
  let testedBaseElement;
  let testedGetBeTestId;
  beforeEach(() => {
    const { baseElement, getByTestId } = render(
      <AutoComplete options={[{ login: 'exmaple' }]} optionKey="login" />
    );
    testedBaseElement = baseElement;
    testedGetBeTestId = getByTestId;
  });
  it('should match the snapshot', () => {
    expect(testedBaseElement).toMatchSnapshot();
  });
  it('should render autoComplete', () => {
    expect(testedGetBeTestId('autoComplete')).toBeInTheDocument();
  });
  it('should render textField', () => {
    expect(testedGetBeTestId('autoComplete__textField')).toBeInTheDocument();
  });
});
