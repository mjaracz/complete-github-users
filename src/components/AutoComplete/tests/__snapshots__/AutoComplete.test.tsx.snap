// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<AutoComplete/> should match the snapshot 1`] = `
<body>
  <div>
    <div
      class="autoComplete"
      data-testid="autoComplete"
    >
      <label
        class="autoComplete__textField"
        data-testid="autoComplete__textField"
      >
        <input
          placeholder=" "
          value=""
        />
        <span>
          provide username
        </span>
      </label>
      <ul
        class="autoComplete__optionsList"
      >
        <li
          class="optionsList__item"
        >
          exmaple
        </li>
      </ul>
    </div>
  </div>
</body>
`;
